python/ray/data/_internal/execution/operators/map_operator_state.py

```diff
@@ -0,0 +1,506 @@
+from dataclasses import dataclass
+from typing import Callable, Optional, List, Dict, Any, Union, Tuple, Iterator
+
+import ray
+from ray.data._internal.compute import (
+    ComputeStrategy,
+    TaskPoolStrategy,
+    ActorPoolStrategy,
+)
+from ray.data._internal.execution.util import merge_ref_bundles
+from ray.data._internal.execution.interfaces import (
+    RefBundle,
+)
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.memory_tracing import trace_allocation
+from ray.data._internal.remote_fn import cached_remote_fn
+from ray.data.block import Block, BlockAccessor, BlockMetadata, BlockExecStats
+from ray.data.context import DEFAULT_SCHEDULING_STRATEGY, DatasetContext
+from ray.types import ObjectRef
+from ray._raylet import ObjectRefGenerator
+
+
+class MapOperatorState:
+    def __init__(
+        self,
+        transform_fn: Callable[[Iterator[Block]], Iterator[Block]],
+        compute_strategy: ComputeStrategy,
+        ray_remote_args: Optional[Dict[str, Any]],
+        min_rows_per_bundle: Optional[int],
+    ):
+        # Execution arguments.
+        self._min_rows_per_bundle: Optional[int] = min_rows_per_bundle
+
+        # Submitter of Ray tasks mapping transform_fn over data.
+        if ray_remote_args is None:
+            ray_remote_args = {}
+        if isinstance(compute_strategy, TaskPoolStrategy):
+            task_submitter = TaskPoolSubmitter(ray_remote_args)
+        elif isinstance(compute_strategy, ActorPoolStrategy):
+            task_submitter = ActorPoolSubmitter(compute_strategy, ray_remote_args)
+        else:
+            raise ValueError(f"Unsupported execution strategy {compute_strategy}")
+        self._task_submitter: MapTaskSubmitter = task_submitter
+        # Whether we have started the task submitter yet.
+        self._have_started_submitter = False
+
+        # Put the function def in the object store to avoid repeated serialization
+        # in case it's large (i.e., closure captures large objects).
+        self._transform_fn_ref = ray.put(transform_fn)
+
+        # The temporary block bundle used to accumulate inputs until they meet the
+        # min_rows_per_bundle requirement.
+        self._block_bundle: Optional[RefBundle] = None
+
+        # Execution state.
+        self._tasks: Dict[ObjectRef[Union[ObjectRefGenerator, Block]], _TaskState] = {}
+        self._tasks_by_output_order: Dict[int, _TaskState] = {}
+        self._next_task_index: int = 0
+        self._next_output_index: int = 0
+        self._obj_store_mem_alloc: int = 0
+        self._obj_store_mem_freed: int = 0
+        self._obj_store_mem_cur: int = 0
+        self._obj_store_mem_peak: int = 0
+
+    def add_input(self, bundle: RefBundle) -> None:
+        if not self._have_started_submitter:
+            # Start the task submitter on the first input.
+            self._task_submitter.start()
+            self._have_started_submitter = True
```

Comment:
This could be cleaner if we had a `PhysicalOperator.start()` or `PhysicalOperator.initialize()` API that the executors used to signal to the operator that execution is actually starting, which might be generically useful since `PhysicalOperator` construction (which happens at the planning stage) can happen far before `PhysicalOperator` execution actually starts. But for now, we just use the first arriving input as the signal.